pkg/lightning/backend/external/engine_test.go

```diff
@@ -396,17 +396,30 @@ func TestEngineOnDup(t *testing.T) {
 		)
 	}
 
-	t.Run("on duplicate ignore and error", func(t *testing.T) {
-		for _, od := range []engineapi.OnDuplicateKey{engineapi.OnDuplicateKeyIgnore, engineapi.OnDuplicateKeyError} {
-			store := storage.NewMemStorage()
-			dataFiles, statFiles := prepareKVFiles(t, store, contents)
-			extEngine := getEngineFn(store, od, dataFiles, statFiles)
-			loadDataCh := make(chan engineapi.DataAndRanges, 4)
-			require.ErrorContains(t, extEngine.LoadIngestData(ctx, loadDataCh), "duplicate key found")
-			t.Cleanup(func() {
-				require.NoError(t, extEngine.Close())
-			})
-		}
+	t.Run("on duplicate ignore", func(t *testing.T) {
+		onDup := engineapi.OnDuplicateKeyIgnore
+		store := storage.NewMemStorage()
+		dataFiles, statFiles := prepareKVFiles(t, store, contents)
+		extEngine := getEngineFn(store, onDup, dataFiles, statFiles)
+		loadDataCh := make(chan engineapi.DataAndRanges, 4)
+		require.ErrorContains(t, extEngine.LoadIngestData(ctx, loadDataCh), "duplicate key found")
+		t.Cleanup(func() {
+			require.NoError(t, extEngine.Close())
+		})
+	})
+
+	t.Run("on duplicate ignore", func(t *testing.T) {
+		// OnDuplicateKeyError is used for adding unique index.
+		// Current implementation would report error when writting to tikv, instead of in LoadIngestData
+		onDup := engineapi.OnDuplicateKeyError
+		store := storage.NewMemStorage()
+		dataFiles, statFiles := prepareKVFiles(t, store, contents)
+		extEngine := getEngineFn(store, onDup, dataFiles, statFiles)
+		loadDataCh := make(chan engineapi.DataAndRanges, 4)
+		require.NoError(t, extEngine.LoadIngestData(ctx, loadDataCh))
```

Comment:
we should return error for OnDuplicateKeyError

Reply:
[`48086f2` (#60848)](https://github.com/pingcap/tidb/pull/60848/commits/48086f220c796a84772eb1469969d03c1a4c1d12)